### PR TITLE
[SC-77396] Block confusing SQL CREATE TABLE behavior

### PR DESCRIPTION
**Pull request by FX196**
Added logic and unit tests to block queries like ``CREATE TABLE delta.`/foo` USING delta LOCATION "/bar"`` where ambiguous paths are supplied.
Users can allow such queries by setting the `DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS` flag to true, in which case the `USING delta LOCATION "/bar"` statement will be ignored.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala [hunk 1]
```diff
@@ -414,6 +414,20 @@ trait DeltaSQLConfBase {
              |""".stripMargin)
       .booleanConf
       .createWithDefault(false)
+
+  val DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS =
+    buildConf("legacy.allowAmbiguousPaths")
+      .internal()
+      .doc("""
+             |Delta was unintentionally allowing CREATE TABLE queries with both 'delta.`path`'
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:422]: Delta was unintentionally allowing CREATE TABLE queries with both 'delta.`path`' and `LOCATION path` clauses. In the new version, we will raise an error for this case. This flag is added to allow users to skip the check. when it's set to true and there are two paths in CREATE TABLE, the `LOCATION path` clause will be ignored like what the old version does.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala [hunk 2]
```diff
@@ -476,4 +475,40 @@ abstract class DeltaDDLTestBase extends QueryTest with SQLTestUtils {
     }
   }
 
+  test("block table creation with ambiguous delta paths") {
```
Comment (Yaohua628) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala:478]: nit: maybe you could put those 2 tests into one block:
```
Seq(true, false).foreach { enabled =>
  withSQLConf(DeltaSQLConf.DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS.key -> enabled.toString)
```
And then a separate one liner test to check DeltaSQLConf.DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS.key is set to false by default.
```
assert(!spark.conf.get(DeltaSQLConf.DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS))
```

**Review by zsxwing — CHANGES_REQUESTED** [review]
Thanks! Most of my comment are code style comments. Could you also add a test to test the behavior when two paths are the same?

**Comment by FX196** [comment]
Style issues addressed and added checks for the default value and the case where two paths are the same.

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaErrors.scala [hunk 3]
```diff
@@ -1146,6 +1146,15 @@ object DeltaErrors
       " cannot be set at the same time. Please set only one group of them.")
   }
 
+  def ambiguousPathsInCreateTableException(identPath: String, propertiesPath: String): Throwable = {
+    new AnalysisException(
+      s"""
+         |CREATE TABLE contains two different locations: ${identPath} and ${propertiesPath}.
+         |You can remove the LOCATION clause from the CREATE TABLE statement, or set
+         |DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS.key to true to skip this check.
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaErrors.scala:1154]: `DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS.key` -> `${DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS.key}`

File: core/src/main/scala/org/apache/spark/sql/delta/DeltaErrors.scala [hunk 4]
```diff
@@ -1146,6 +1146,15 @@ object DeltaErrors
       " cannot be set at the same time. Please set only one group of them.")
   }
 
+  def ambiguousPathsInCreateTableException(identPath: String, propertiesPath: String): Throwable = {
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/sql/delta/DeltaErrors.scala:1149]: nit: `identPath` -> `identifier`, `propertiesPath` -> `location`.

File: core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala [hunk 5]
```diff
@@ -414,6 +414,18 @@ trait DeltaSQLConfBase {
              |""".stripMargin)
       .booleanConf
       .createWithDefault(false)
+
+  val DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS =
+    buildConf("legacy.allowAmbiguousPathsInCreateTable")
+      .internal()
+      .doc("""
+             |Delta was unintentionally allowing CREATE TABLE queries with both 'delta.`path`'
+             |and 'USING delta LOCATION path' clauses. In the new version, we will raise an error
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSQLConf.scala:423]: nit: `USING delta` is redundant.

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala [hunk 6]
```diff
@@ -476,4 +473,47 @@ abstract class DeltaDDLTestBase extends QueryTest with SQLTestUtils {
...
+        sql(s"CREATE TABLE delta.`$fooPath`(id LONG) USING delta LOCATION '$fooPath'")
+        assert(io.delta.tables.DeltaTable.isDeltaTable(fooPath))
+      }
+    }
+
+    // make sure default is false
+    assert(!spark.conf.get(DeltaSQLConf.DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS))
+  }
+
+  test("allow ambiguous delta paths with legacy flag") {
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala:515]: forgot to remove?

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala [hunk 7]
```diff
@@ -476,4 +473,47 @@ abstract class DeltaDDLTestBase extends QueryTest with SQLTestUtils {
...
+          assert(io.delta.tables.DeltaTable.isDeltaTable(fooPath))
+          assert(!io.delta.tables.DeltaTable.isDeltaTable(barPath))
+        }
+      }
+    }
+
+    // allowed if paths are the same
+    withTempDir { foo =>
+      val fooPath = foo.getCanonicalPath()
+      withSQLConf(DeltaSQLConf.DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS.key -> "true") {
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala:505]: nit: this should be removed

File: core/src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala [hunk 8]
```diff
@@ -476,4 +473,47 @@ abstract class DeltaDDLTestBase extends QueryTest with SQLTestUtils {
...
+        sql(s"CREATE TABLE delta.`$fooPath`(id LONG) USING delta LOCATION '$fooPath'")
+        assert(io.delta.tables.DeltaTable.isDeltaTable(fooPath))
+      }
+    }
+
+    // make sure default is false
+    assert(!spark.conf.get(DeltaSQLConf.DELTA_LEGACY_ALLOW_AMBIGUOUS_PATHS))
```
Comment (zsxwing) [inline core/src/test/scala/org/apache/spark/sql/delta/DeltaDDLSuite.scala:512]: we don't need this. The test that verifies the behavior without setting the flag has tested the default value.

**Review by zsxwing — CHANGES_REQUESTED** [review]

**Review by zsxwing — APPROVED** [review]
LGTM. Thanks!

**Comment by FX196** [comment]
Imported

**Pull request closed without merging**